experiments/kompanion/cmd/migrator/migrator.go

```diff
@@ -0,0 +1,369 @@
+type Task interface {
+	Run(ctx context.Context) error
+}
+
+// tracks the namespaces to be exported.
+// thread safe.
+type taskQueue struct {
+	mu    sync.Mutex
+	tasks []Task // will be treated as a FIFO queue
+}
+
+func (n *taskQueue) GetWork() Task {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	if len(n.tasks) == 0 {
+		return nil
+	}
+
+	workItem := n.tasks[0]
+	n.tasks = n.tasks[1:]
+
+	return workItem
+}
+
+func (n *taskQueue) AddTask(t Task) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	n.tasks = append(n.tasks, t)
+}
+
+type dumpResourcesTask struct {
+	// Namespace is the namespace to filter down
+	Namespace string
+
+	DynamicClient *dynamic.DynamicClient
+
+	OutputChannel chan string
+
+	// Resources is the list of resources to query
+	Resources []schema.GroupVersionResource
+
+	ShouldExcludeObject func(id types.NamespacedName) bool
+}
+
+func (t *dumpResourcesTask) Run(ctx context.Context) error {
+	for _, gvr := range t.Resources {
+		var resources *unstructured.UnstructuredList
+		if t.Namespace != "" {
+			r, err := t.DynamicClient.Resource(gvr).Namespace(t.Namespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("fetching gvr %s resources in namespace %s: %w", gvr, t.Namespace, err)
+			}
+			resources = r
+
+		} else {
+			r, err := t.DynamicClient.Resource(gvr).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("fetching gvr %s resources: %w", gvr, err)
+			}
+			resources = r
+		}
+
+		for _, r := range resources.Items {
+			id := types.NamespacedName{
+				Namespace: r.GetNamespace(),
+				Name:      r.GetName(),
+			}
+			if t.ShouldExcludeObject(id) {
+				continue
+			}
+			r = pruneResource(r)
+
+			data, err := yaml.Marshal(r)
+			if err != nil {
+				return fmt.Errorf("error marshalling resource %s: %w", id, err)
+			}
+			t.OutputChannel <- string(data)
+		}
+	}
+
+	return nil
+}
+
+func pruneResource(r unstructured.Unstructured) unstructured.Unstructured {
+	newAnnotations := make(map[string]string)
+	annotations := r.GetAnnotations()
+	for key, value := range annotations {
+		if !strings.Contains(key, "cnrm.cloud.google.com") {
+			continue
+		}
+		newAnnotations[key] = value
+	}
+	if len(newAnnotations) == 0 {
+		r.SetAnnotations(nil)
+	} else {
+		r.SetAnnotations(newAnnotations)
+	}
+
+	r.SetCreationTimestamp(metav1.Time{})
+	r.SetFinalizers(nil)
+	r.SetGeneration(0)
+	r.SetManagedFields(nil)
+	r.SetResourceVersion("")
+	r.SetUID("")
+
+	unstructured.RemoveNestedField(r.Object, "status", "conditions")
+	unstructured.RemoveNestedField(r.Object, "status", "creationTimestamp")
+	unstructured.RemoveNestedField(r.Object, "status", "healthy")
+	unstructured.RemoveNestedField(r.Object, "status", "observedGeneration")
+	unstructured.RemoveNestedField(r.Object, "status", "selfLink") // Are we sure?
+	status, present, err := unstructured.NestedMap(r.Object, "status")
+	if err != nil {
+		log.Fatalf("Could not get the status field %v\r", err)
```

Comment:
This will kill the goroutine right? Do we want to just Log the err instead or return it out of this closure?